File: model/src/main/scala/ru/mipt/acsl/decode/model/proxy/ResolvingResult.scala
package ru.mipt.acsl.decode.model.proxy

/**
  * Created by metadeus on 27.05.16.
  */
trait ResolvingResult[T] {

  def result: Option[T]

  def messages: ResolvingMessages

  def messages_=(messages: ResolvingMessages)

}

object ResolvingResult {

  private case class ResolvingResultImpl[T](var result: Option[T], var messages: ResolvingMessages) extends ResolvingResult[T]

  def empty[T]: ResolvingResult[T] = ResolvingResultImpl(None, ResolvingMessages.empty)

  def apply[T](result: Option[T] = None, messages: ResolvingMessages = ResolvingMessages.empty): ResolvingResult[T] =
    ResolvingResultImpl(result, messages)

}

File: model/src/main/java/ru/mipt/acsl/decode/model/proxy/ResolvingResult.java
package ru.mipt.acsl.decode.model.proxy;

import org.jetbrains.annotations.Nullable;
import ru.mipt.acsl.decode.model.Message;

import java.util.Optional;

/**
 * @author Artem Shein
 */
public interface ResolvingResult<T> {

    static <T> ResolvingResult<T> newInstance() {
        return newInstance(null);
    }

    static <T> ResolvingResult<T> newInstance(Message msg) {
        return newInstance(null, ResolvingMessages.newInstance(msg));
    }

    static <T> ResolvingResult<T> newInstance(@Nullable T result) {
        return newInstance(result, ResolvingMessages.newInstance());
    }

    static <T> ResolvingResult<T> newInstance(@Nullable T result, Message message) {
        return newInstance(result, ResolvingMessages.newInstance(message));
    }

    static <T> ResolvingResult<T> newInstance(@Nullable T result, ResolvingMessages messages) {
        return new ResolvingResultImpl<>(result, messages);
    }

    Optional<T> result();

    ResolvingMessages messages();

    void messages(ResolvingMessages messages);

}
